table/tables/tables.go

```diff
@@ -187,115 +213,100 @@ func (t *Table) FirstKey() kv.Key {
-	// We should check whether this table has on update column which state is write only.
-	currentData := make([]types.Datum, len(t.WritableCols()))
-	copy(currentData, newData)
+func (t *Table) UpdateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, touched []bool) error {
+	txn := ctx.Txn()
+	bs := kv.NewBufferStore(txn)
 
-	// If they are not set, and other data are changed, they will be updated by current timestamp too.
-	err := t.setOnUpdateData(ctx, touched, currentData)
+	// rebuild index
+	err := t.rebuildIndices(bs, h, touched, oldData, newData)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	txn := ctx.Txn()
-	bs := kv.NewBufferStore(ctx.Txn())
-
-	// Compose new row
-	t.composeNewData(touched, currentData, oldData)
-	colIDs := make([]int64, 0, len(t.WritableCols()))
-	for i, col := range t.WritableCols() {
-		if col.State != model.StatePublic && currentData[i].IsNull() {
-			defaultVal, err1 := table.GetColDefaultValue(ctx, col.ToInfo())
-			if err1 != nil {
-				return errors.Trace(err1)
+	var colIDs, binlogColIDs []int64
+	var row, binlogOldRow, binlogNewRow []types.Datum
+	colIDs = make([]int64, 0, len(newData))
+	row = make([]types.Datum, 0, len(newData))
+	if shouldWriteBinlog(ctx) {
+		binlogColIDs = make([]int64, 0, len(newData))
+		binlogOldRow = make([]types.Datum, 0, len(newData))
+		binlogNewRow = make([]types.Datum, 0, len(newData))
+	}
+
+	for _, col := range t.WritableCols() {
+		var value types.Datum
+		if col.State != model.StatePublic {
+			// if col is in write only or write reorganization state
+			// and the value is not default, we should use default value.
+			value, err = table.GetColDefaultValue(ctx, col.ToInfo())
+			if err != nil {
+				return errors.Trace(err)
+			}
+			cmp, err := oldData[col.Offset].CompareDatum(ctx.GetSessionVars().StmtCtx, value)
+			if err != nil {
+				return errors.Trace(err)
 			}
-			currentData[i] = defaultVal
+			if cmp != 0 {
+				value = oldData[col.Offset]
+			}
+		} else {
+			value = newData[col.Offset]
+		}
+		if !t.canSkip(col, value) {
+			colIDs = append(colIDs, col.ID)
+			row = append(row, value)
+		}
+		if shouldWriteBinlog(ctx) && !t.canSkipUpdateBinlog(col, value) {
+			binlogColIDs = append(binlogColIDs, col.ID)
+			binlogOldRow = append(binlogOldRow, oldData[col.Offset])
+			binlogNewRow = append(binlogNewRow, value)
 		}
-		colIDs = append(colIDs, col.ID)
 	}
-	// Set new row data into KV.
+
 	key := t.RecordKey(h)
-	value, err := tablecodec.EncodeRow(currentData, colIDs, ctx.GetSessionVars().GetTimeZone())
+	value, err := tablecodec.EncodeRow(row, colIDs, ctx.GetSessionVars().GetTimeZone())
 	if err != nil {
 		return errors.Trace(err)
 	}
 	if err = bs.Set(key, value); err != nil {
 		return errors.Trace(err)
 	}
-
-	// rebuild index
-	if err = t.rebuildIndices(bs, h, touched, oldData, currentData); err != nil {
```

Comment:
If error happens, all operations will rollback, because they are in same transaction.
Rebuild index first because we `buildIndices` in `AddRecord` first. I think make them same is better.
The for loop below this just chooses which columns should be written into TiKV and which columns should be written into binlog. It won't change the row.